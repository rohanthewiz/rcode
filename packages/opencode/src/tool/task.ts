import { Tool } from "./tool"
import DESCRIPTION from "./task.txt"
import { z } from "zod"
import { Session } from "../session"
import { Bus } from "../bus"
import { Message } from "../session/message"

export const TaskTool = Tool.define({
  id: "task",
  description: DESCRIPTION,
  parameters: z.object({
    description: z
      .string()
      .describe("A short (3-5 words) description of the task"),
    prompt: z.string().describe("The task for the agent to perform"),
  }),
  async execute(params, ctx) {
    const session = await Session.create(ctx.sessionID)
    const msg = await Session.getMessage(ctx.sessionID, ctx.messageID)
    const metadata = msg.metadata.assistant!

    function summary(input: Message.Info) {
      const result = []

      for (const part of input.parts) {
        if (part.type === "tool-invocation") {
          result.push({
            toolInvocation: part.toolInvocation,
            metadata: input.metadata.tool[part.toolInvocation.toolCallId],
          })
        }
      }
      return result
    }

    const unsub = Bus.subscribe(Message.Event.Updated, async (evt) => {
      if (evt.properties.info.metadata.sessionID !== session.id) return
      ctx.metadata({
        title: params.description,
        summary: summary(evt.properties.info),
      })
    })

    ctx.abort.addEventListener("abort", () => {
      Session.abort(session.id)
    })
    const result = await Session.chat({
      sessionID: session.id,
      modelID: metadata.modelID,
      providerID: metadata.providerID,
      parts: [
        {
          type: "text",
          text: params.prompt,
        },
      ],
    })
    unsub()
    return {
      metadata: {
        title: params.description,
        summary: summary(result),
      },
      output: result.parts.findLast((x) => x.type === "text")!.text,
    }
  },
})
