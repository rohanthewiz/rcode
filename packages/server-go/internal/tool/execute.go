package tool

import (
	"context"
	"fmt"
	"time"

	"github.com/rohanthewiz/logger"
	"github.com/rohanthewiz/serr"
)

// Executor handles the execution of tools within the context of a message.
// It manages parameter validation, timeout handling, and error recovery.
type Executor struct {
	registry *Registry
}

// NewExecutor creates a new tool executor
func NewExecutor(registry *Registry) *Executor {
	return &Executor{
		registry: registry,
	}
}

// Execute runs a tool with the given parameters.
// It handles validation, timeout, and error handling.
func (e *Executor) Execute(toolID string, params map[string]any, ctx Context) (Result, error) {
	// Retrieve the tool from registry
	tool, exists := e.registry.Get(toolID)
	if !exists {
		return Result{}, serr.New("tool not found: %s", toolID)
	}
	
	// Validate parameters against schema
	if err := tool.Parameters().Validate(params); err != nil {
		return Result{}, serr.Wrap(err, "parameter validation failed")
	}
	
	// Create execution context with timeout
	execCtx, cancel := context.WithCancel(ctx.Abort)
	defer cancel()
	
	// Track execution time
	start := time.Now()
	
	// Channel for collecting results
	resultChan := make(chan Result, 1)
	errChan := make(chan error, 1)
	
	// Execute tool in goroutine to handle cancellation
	go func() {
		defer func() {
			// Recover from panics in tool execution
			if r := recover(); r != nil {
				errChan <- serr.New("tool panicked: %v", r)
			}
		}()
		
		// Execute the tool
		result, err := tool.Execute(Context{
			SessionID: ctx.SessionID,
			MessageID: ctx.MessageID,
			Abort:     execCtx,
			Metadata:  ctx.Metadata,
		}, params)
		
		if err != nil {
			errChan <- err
		} else {
			resultChan <- result
		}
	}()
	
	// Wait for result or cancellation
	select {
	case result := <-resultChan:
		duration := time.Since(start)
		logger.Info("tool executed successfully",
			"tool", toolID,
			"duration", duration,
			"session", ctx.SessionID,
		)
		
		// Add execution metadata
		if result.Metadata == nil {
			result.Metadata = make(map[string]any)
		}
		result.Metadata["duration_ms"] = duration.Milliseconds()
		
		return result, nil
		
	case err := <-errChan:
		duration := time.Since(start)
		logger.LogErr(err, "tool execution failed",
			"tool", toolID,
			"duration", duration,
			"session", ctx.SessionID,
		)
		
		// Return error as output for AI to see
		return Result{
			Output: fmt.Sprintf("Error: %v", err),
			Metadata: map[string]any{
				"error":       err.Error(),
				"duration_ms": duration.Milliseconds(),
			},
		}, nil
		
	case <-execCtx.Done():
		// Execution was cancelled
		logger.Info("tool execution cancelled",
			"tool", toolID,
			"session", ctx.SessionID,
		)
		
		return Result{
			Output: "Execution cancelled",
			Metadata: map[string]any{
				"cancelled": true,
			},
		}, nil
	}
}

// ExecuteMultiple runs multiple tools concurrently.
// This is useful when the AI requests multiple tool executions at once.
func (e *Executor) ExecuteMultiple(executions []ToolExecution, ctx Context) []ExecutionResult {
	results := make([]ExecutionResult, len(executions))
	var wg sync.WaitGroup
	
	for i, exec := range executions {
		wg.Add(1)
		go func(idx int, execution ToolExecution) {
			defer wg.Done()
			
			// Create a unique metadata callback for this execution
			metadataCallback := func(meta map[string]any) {
				// Include tool call ID in metadata updates
				meta["tool_call_id"] = execution.CallID
				ctx.Metadata(meta)
			}
			
			// Execute the tool
			result, err := e.Execute(
				execution.ToolID,
				execution.Parameters,
				Context{
					SessionID: ctx.SessionID,
					MessageID: ctx.MessageID,
					Abort:     ctx.Abort,
					Metadata:  metadataCallback,
				},
			)
			
			results[idx] = ExecutionResult{
				CallID: execution.CallID,
				Result: result,
				Error:  err,
			}
		}(i, exec)
	}
	
	wg.Wait()
	return results
}

// ToolExecution represents a single tool execution request
type ToolExecution struct {
	CallID     string
	ToolID     string
	Parameters map[string]any
}

// ExecutionResult contains the result of a tool execution
type ExecutionResult struct {
	CallID string
	Result Result
	Error  error
}