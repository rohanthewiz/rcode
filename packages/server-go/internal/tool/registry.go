package tool

import (
	"fmt"
	"sync"
)

// Registry manages available tools
type Registry struct {
	tools map[string]Tool
	mu    sync.RWMutex
}

// NewRegistry creates a new tool registry
func NewRegistry() *Registry {
	return &Registry{
		tools: make(map[string]Tool),
	}
}

// Register adds a tool to the registry
func (r *Registry) Register(tool Tool) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if tool == nil {
		return fmt.Errorf("cannot register nil tool")
	}

	id := tool.ID()
	if id == "" {
		return fmt.Errorf("tool has empty ID")
	}

	if _, exists := r.tools[id]; exists {
		return fmt.Errorf("tool with ID %q already registered", id)
	}

	r.tools[id] = tool
	return nil
}

// Get retrieves a tool by ID
func (r *Registry) Get(id string) Tool {
	r.mu.RLock()
	defer r.mu.RUnlock()

	return r.tools[id]
}

// GetAll returns all registered tools
func (r *Registry) GetAll() []Tool {
	r.mu.RLock()
	defer r.mu.RUnlock()

	tools := make([]Tool, 0, len(r.tools))
	for _, tool := range r.tools {
		tools = append(tools, tool)
	}
	return tools
}

// GetFiltered returns tools that pass the filter function
func (r *Registry) GetFiltered(filter func(Tool) bool) []Tool {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var filtered []Tool
	for _, tool := range r.tools {
		if filter(tool) {
			filtered = append(filtered, tool)
		}
	}
	return filtered
}

// Has checks if a tool with the given ID exists
func (r *Registry) Has(id string) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()

	_, exists := r.tools[id]
	return exists
}

// Remove removes a tool from the registry
func (r *Registry) Remove(id string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.tools[id]; exists {
		delete(r.tools, id)
		return true
	}
	return false
}

// Count returns the number of registered tools
func (r *Registry) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()

	return len(r.tools)
}

// Clear removes all tools from the registry
func (r *Registry) Clear() {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.tools = make(map[string]Tool)
}