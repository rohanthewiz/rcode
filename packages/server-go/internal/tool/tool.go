package tool

import (
	"context"
	"sync"

	"github.com/rohanthewiz/serr"
)

// Tool represents a capability that can be invoked by the AI assistant.
// Each tool has a unique ID, description, parameter schema, and execution logic.
type Tool interface {
	// ID returns the unique identifier for this tool
	ID() string
	
	// Description returns a detailed description of what this tool does
	Description() string
	
	// Parameters returns the schema for validating input parameters
	Parameters() Schema
	
	// Execute runs the tool with the given parameters and context
	Execute(ctx Context, params map[string]any) (Result, error)
}

// Context provides runtime information for tool execution.
// It includes session/message identifiers, cancellation support, and metadata updates.
type Context struct {
	// SessionID identifies the current chat session
	SessionID string
	
	// MessageID identifies the message that triggered this tool execution
	MessageID string
	
	// Abort provides cancellation support via context
	Abort context.Context
	
	// Metadata allows tools to send real-time updates during execution
	Metadata func(meta map[string]any)
}

// Result contains the output from a tool execution.
// Output is the main text result, while Metadata provides structured data.
type Result struct {
	// Output is the text result that will be shown to the user
	Output string
	
	// Metadata contains structured data about the execution
	Metadata map[string]any
}

// Registry manages the collection of available tools.
// It provides thread-safe registration and retrieval of tools.
type Registry struct {
	tools map[string]Tool
	mu    sync.RWMutex
}

// NewRegistry creates a new tool registry
func NewRegistry() *Registry {
	return &Registry{
		tools: make(map[string]Tool),
	}
}

// Register adds a tool to the registry.
// Returns an error if a tool with the same ID already exists.
func (r *Registry) Register(tool Tool) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	
	if _, exists := r.tools[tool.ID()]; exists {
		return serr.New("tool with ID %s already registered", tool.ID())
	}
	
	r.tools[tool.ID()] = tool
	return nil
}

// Get retrieves a tool by ID.
// Returns the tool and true if found, nil and false otherwise.
func (r *Registry) Get(id string) (Tool, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	
	tool, ok := r.tools[id]
	return tool, ok
}

// List returns all registered tools
func (r *Registry) List() []Tool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	
	tools := make([]Tool, 0, len(r.tools))
	for _, tool := range r.tools {
		tools = append(tools, tool)
	}
	return tools
}

// ForProvider returns tools available for a specific AI provider.
// This allows customizing which tools are available for different providers.
func (r *Registry) ForProvider(provider string) []Tool {
	// For now, return all tools. In the future, we can add provider-specific filtering
	// based on tool capabilities and provider limitations.
	return r.List()
}

// Schema represents a parameter validation schema.
// It defines the structure and constraints for tool input parameters.
type Schema interface {
	// Validate checks if the given value conforms to this schema
	Validate(value any) error
	
	// Description returns a human-readable description of this schema
	Description() string
	
	// ToJSON converts the schema to a JSON-compatible representation
	ToJSON() map[string]any
}