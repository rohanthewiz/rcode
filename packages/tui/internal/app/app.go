package app

import (
	"context"
	"fmt"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"log/slog"

	tea "github.com/charmbracelet/bubbletea/v2"
	"github.com/sst/opencode-sdk-go"
	"github.com/sst/opencode/internal/commands"
	"github.com/sst/opencode/internal/components/toast"
	"github.com/sst/opencode/internal/config"
	"github.com/sst/opencode/internal/styles"
	"github.com/sst/opencode/internal/theme"
	"github.com/sst/opencode/internal/util"
)

var RootPath string
var CwdPath string

type App struct {
	Info      opencode.App
	Version   string
	StatePath string
	Config    *opencode.Config
	Client    *opencode.Client
	State     *config.State
	Provider  *opencode.Provider
	Model     *opencode.Model
	Session   *opencode.Session
	Messages  []opencode.Message
	Commands  commands.CommandRegistry
}

type SessionSelectedMsg = *opencode.Session
type ModelSelectedMsg struct {
	Provider opencode.Provider
	Model    opencode.Model
}
type SessionClearedMsg struct{}
type CompactSessionMsg struct{}
type SendMsg struct {
	Text        string
	Attachments []Attachment
}
type CompletionDialogTriggeredMsg struct {
	InitialValue string
}
type OptimisticMessageAddedMsg struct {
	Message opencode.Message
}

func New(
	ctx context.Context,
	version string,
	appInfo opencode.App,
	httpClient *opencode.Client,
) (*App, error) {
	RootPath = appInfo.Path.Root
	CwdPath = appInfo.Path.Cwd

	configInfo, err := httpClient.Config.Get(ctx)
	if err != nil {
		return nil, err
	}

	if configInfo.Keybinds.Leader == "" {
		configInfo.Keybinds.Leader = "ctrl+x"
	}

	appStatePath := filepath.Join(appInfo.Path.State, "tui")
	appState, err := config.LoadState(appStatePath)
	if err != nil {
		appState = config.NewState()
		config.SaveState(appStatePath, appState)
	}

	if configInfo.Theme != "" {
		appState.Theme = configInfo.Theme
	}

	if configInfo.Model != "" {
		splits := strings.Split(configInfo.Model, "/")
		appState.Provider = splits[0]
		appState.Model = strings.Join(splits[1:], "/")
	}

	if err := theme.LoadThemesFromDirectories(
		appInfo.Path.Config,
		appInfo.Path.Root,
		appInfo.Path.Cwd,
	); err != nil {
		slog.Warn("Failed to load themes from directories", "error", err)
	}

	if appState.Theme != "" {
		if appState.Theme == "system" && styles.Terminal != nil {
			theme.UpdateSystemTheme(
				styles.Terminal.Background,
				styles.Terminal.BackgroundIsDark,
			)
		}
		theme.SetTheme(appState.Theme)
	}

	slog.Debug("Loaded config", "config", configInfo)

	app := &App{
		Info:      appInfo,
		Version:   version,
		StatePath: appStatePath,
		Config:    configInfo,
		State:     appState,
		Client:    httpClient,
		Session:   &opencode.Session{},
		Messages:  []opencode.Message{},
		Commands:  commands.LoadFromConfig(configInfo),
	}

	return app, nil
}

func (a *App) InitializeProvider() tea.Cmd {
	return func() tea.Msg {
		providersResponse, err := a.Client.Config.Providers(context.Background())
		if err != nil {
			slog.Error("Failed to list providers", "error", err)
			// TODO: notify user
			return nil
		}
		providers := providersResponse.Providers
		var defaultProvider *opencode.Provider
		var defaultModel *opencode.Model

		var anthropic *opencode.Provider
		for _, provider := range providers {
			if provider.ID == "anthropic" {
				anthropic = &provider
			}
		}

		// default to anthropic if available
		if anthropic != nil {
			defaultProvider = anthropic
			defaultModel = getDefaultModel(providersResponse, *anthropic)
		}

		for _, provider := range providers {
			if defaultProvider == nil || defaultModel == nil {
				defaultProvider = &provider
				defaultModel = getDefaultModel(providersResponse, provider)
			}
			providers = append(providers, provider)
		}
		if len(providers) == 0 {
			slog.Error("No providers configured")
			return nil
		}

		var currentProvider *opencode.Provider
		var currentModel *opencode.Model
		for _, provider := range providers {
			if provider.ID == a.State.Provider {
				currentProvider = &provider

				for _, model := range provider.Models {
					if model.ID == a.State.Model {
						currentModel = &model
					}
				}
			}
		}
		if currentProvider == nil || currentModel == nil {
			currentProvider = defaultProvider
			currentModel = defaultModel
		}

		return ModelSelectedMsg{
			Provider: *currentProvider,
			Model:    *currentModel,
		}
	}
}

func getDefaultModel(response *opencode.ConfigProvidersResponse, provider opencode.Provider) *opencode.Model {
	if match, ok := response.Default[provider.ID]; ok {
		model := provider.Models[match]
		return &model
	} else {
		for _, model := range provider.Models {
			return &model
		}
	}
	return nil
}

type Attachment struct {
	FilePath string
	FileName string
	MimeType string
	Content  []byte
}

func (a *App) IsBusy() bool {
	if len(a.Messages) == 0 {
		return false
	}

	lastMessage := a.Messages[len(a.Messages)-1]
	return lastMessage.Metadata.Time.Completed == 0
}

func (a *App) SaveState() {
	err := config.SaveState(a.StatePath, a.State)
	if err != nil {
		slog.Error("Failed to save state", "error", err)
	}
}

func (a *App) InitializeProject(ctx context.Context) tea.Cmd {
	cmds := []tea.Cmd{}

	session, err := a.CreateSession(ctx)
	if err != nil {
		// status.Error(err.Error())
		return nil
	}

	a.Session = session
	cmds = append(cmds, util.CmdHandler(SessionSelectedMsg(session)))

	go func() {
		_, err := a.Client.Session.Init(ctx, a.Session.ID, opencode.SessionInitParams{
			ProviderID: opencode.F(a.Provider.ID),
			ModelID:    opencode.F(a.Model.ID),
		})
		if err != nil {
			slog.Error("Failed to initialize project", "error", err)
			// status.Error(err.Error())
		}
	}()

	return tea.Batch(cmds...)
}

func (a *App) CompactSession(ctx context.Context) tea.Cmd {
	go func() {
		_, err := a.Client.Session.Summarize(ctx, a.Session.ID, opencode.SessionSummarizeParams{
			ProviderID: opencode.F(a.Provider.ID),
			ModelID:    opencode.F(a.Model.ID),
		})
		if err != nil {
			slog.Error("Failed to compact session", "error", err)
		}
	}()
	return nil
}

func (a *App) MarkProjectInitialized(ctx context.Context) error {
	_, err := a.Client.App.Init(ctx)
	if err != nil {
		slog.Error("Failed to mark project as initialized", "error", err)
		return err
	}
	return nil
}

func (a *App) CreateSession(ctx context.Context) (*opencode.Session, error) {
	session, err := a.Client.Session.New(ctx)
	if err != nil {
		return nil, err
	}
	return session, nil
}

func (a *App) SendChatMessage(ctx context.Context, text string, attachments []Attachment) tea.Cmd {
	var cmds []tea.Cmd
	if a.Session.ID == "" {
		session, err := a.CreateSession(ctx)
		if err != nil {
			return toast.NewErrorToast(err.Error())
		}
		a.Session = session
		cmds = append(cmds, util.CmdHandler(SessionSelectedMsg(session)))
	}

	optimisticMessage := opencode.Message{
		ID:   fmt.Sprintf("optimistic-%d", time.Now().UnixNano()),
		Role: opencode.MessageRoleUser,
		Parts: []opencode.MessagePart{{
			Type: opencode.MessagePartTypeText,
			Text: text,
		}},
		Metadata: opencode.MessageMetadata{
			SessionID: a.Session.ID,
			Time: opencode.MessageMetadataTime{
				Created: float64(time.Now().Unix()),
			},
		},
	}

	a.Messages = append(a.Messages, optimisticMessage)
	cmds = append(cmds, util.CmdHandler(OptimisticMessageAddedMsg{Message: optimisticMessage}))

	cmds = append(cmds, func() tea.Msg {
		_, err := a.Client.Session.Chat(ctx, a.Session.ID, opencode.SessionChatParams{
			Parts: opencode.F([]opencode.MessagePartUnionParam{
				opencode.TextPartParam{
					Type: opencode.F(opencode.TextPartTypeText),
					Text: opencode.F(text),
				},
			}),
			ProviderID: opencode.F(a.Provider.ID),
			ModelID:    opencode.F(a.Model.ID),
		})
		if err != nil {
			errormsg := fmt.Sprintf("failed to send message: %v", err)
			slog.Error(errormsg)
			return toast.NewErrorToast(errormsg)()
		}
		return nil
	})

	// The actual response will come through SSE
	// For now, just return success
	return tea.Batch(cmds...)
}

func (a *App) Cancel(ctx context.Context, sessionID string) error {
	_, err := a.Client.Session.Abort(ctx, sessionID)
	if err != nil {
		slog.Error("Failed to cancel session", "error", err)
		// status.Error(err.Error())
		return err
	}
	return nil
}

func (a *App) ListSessions(ctx context.Context) ([]opencode.Session, error) {
	response, err := a.Client.Session.List(ctx)
	if err != nil {
		return nil, err
	}
	if response == nil {
		return []opencode.Session{}, nil
	}
	sessions := *response
	sort.Slice(sessions, func(i, j int) bool {
		return sessions[i].Time.Created-sessions[j].Time.Created > 0
	})
	return sessions, nil
}

func (a *App) DeleteSession(ctx context.Context, sessionID string) error {
	_, err := a.Client.Session.Delete(ctx, sessionID)
	if err != nil {
		slog.Error("Failed to delete session", "error", err)
		return err
	}
	return nil
}

func (a *App) ListMessages(ctx context.Context, sessionId string) ([]opencode.Message, error) {
	response, err := a.Client.Session.Messages(ctx, sessionId)
	if err != nil {
		return nil, err
	}
	if response == nil {
		return []opencode.Message{}, nil
	}
	messages := *response
	return messages, nil
}

func (a *App) ListProviders(ctx context.Context) ([]opencode.Provider, error) {
	response, err := a.Client.Config.Providers(ctx)
	if err != nil {
		return nil, err
	}
	if response == nil {
		return []opencode.Provider{}, nil
	}

	providers := *response
	return providers.Providers, nil
}

// func (a *App) loadCustomKeybinds() {
//
// }
