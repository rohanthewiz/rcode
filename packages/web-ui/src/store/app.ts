// Global app state management
import { create } from 'zustand';
import type { AppInfo, Provider } from '@/types/api';
import { api } from '@/api/client';

interface AppState {
  // State
  appInfo: AppInfo | null;
  providers: Provider[];
  defaultModels: Record<string, string>;
  isInitialized: boolean;
  isLoading: boolean;
  error: string | null;

  // Actions
  initialize: () => Promise<void>;
  loadProviders: () => Promise<void>;
  setError: (error: string | null) => void;
  clearError: () => void;
}

export const useAppStore = create<AppState>((set) => ({
  // Initial state
  appInfo: null,
  providers: [],
  defaultModels: {},
  isInitialized: false,
  isLoading: false,
  error: null,

  // Actions
  initialize: async () => {
    set({ isLoading: true, error: null });
    try {
      const info = await api.app.info();
      set({ appInfo: info, isInitialized: true });
      
      // Load providers after getting app info
      const { providers, default: defaultModels } = await api.config.providers();
      set({ providers, defaultModels });
    } catch (error) {
      set({ error: error instanceof Error ? error.message : 'Failed to initialize app' });
      throw error;
    } finally {
      set({ isLoading: false });
    }
  },

  loadProviders: async () => {
    try {
      const { providers, default: defaultModels } = await api.config.providers();
      set({ providers, defaultModels });
    } catch (error) {
      set({ error: error instanceof Error ? error.message : 'Failed to load providers' });
      throw error;
    }
  },

  setError: (error) => set({ error }),
  clearError: () => set({ error: null }),
}));