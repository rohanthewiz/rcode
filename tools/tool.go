package tools

import (
	"encoding/json"
)

// Tool represents a tool that can be used by the AI
type Tool struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	InputSchema map[string]interface{} `json:"input_schema"`
}

// ToolUse represents a tool use request from the AI
type ToolUse struct {
	Type  string                 `json:"type"`
	ID    string                 `json:"id"`
	Name  string                 `json:"name"`
	Input map[string]interface{} `json:"input"`
}

// ToolResult represents the result of executing a tool
type ToolResult struct {
	Type      string `json:"type"`
	ToolUseID string `json:"tool_use_id"`
	Content   string `json:"content"`
}

// Executor is the interface for tool execution
type Executor interface {
	Execute(input map[string]interface{}) (string, error)
}

// Registry holds all available tools
type Registry struct {
	tools     map[string]Tool
	executors map[string]Executor
}

// NewRegistry creates a new tool registry
func NewRegistry() *Registry {
	return &Registry{
		tools:     make(map[string]Tool),
		executors: make(map[string]Executor),
	}
}

// Register adds a tool to the registry
func (r *Registry) Register(tool Tool, executor Executor) {
	r.tools[tool.Name] = tool
	r.executors[tool.Name] = executor
}

// GetTools returns all registered tools
func (r *Registry) GetTools() []Tool {
	tools := make([]Tool, 0, len(r.tools))
	for _, tool := range r.tools {
		tools = append(tools, tool)
	}
	return tools
}

// Execute runs a tool and returns the result
func (r *Registry) Execute(toolUse ToolUse) (*ToolResult, error) {
	executor, exists := r.executors[toolUse.Name]
	if !exists {
		return nil, &ToolError{Message: "Unknown tool: " + toolUse.Name}
	}

	result, err := executor.Execute(toolUse.Input)
	if err != nil {
		// Return both the error result and the error itself
		// This allows the enhanced registry to handle retries
		return &ToolResult{
			Type:      "tool_result",
			ToolUseID: toolUse.ID,
			Content:   "Error: " + err.Error(),
		}, err
	}

	return &ToolResult{
		Type:      "tool_result",
		ToolUseID: toolUse.ID,
		Content:   result,
	}, nil
}

// ToolError represents a tool execution error
type ToolError struct {
	Message string
}

func (e *ToolError) Error() string {
	return e.Message
}

// Helper function to get string from interface{}
func GetString(input map[string]interface{}, key string) (string, bool) {
	val, exists := input[key]
	if !exists {
		return "", false
	}
	str, ok := val.(string)
	return str, ok
}

// Helper function to get int from interface{}
func GetInt(input map[string]interface{}, key string) (int, bool) {
	val, exists := input[key]
	if !exists {
		return 0, false
	}

	// Handle both int and float64 (JSON numbers are float64)
	switch v := val.(type) {
	case int:
		return v, true
	case float64:
		return int(v), true
	default:
		return 0, false
	}
}

// Helper function to get bool from interface{}
func GetBool(input map[string]interface{}, key string) (bool, bool) {
	val, exists := input[key]
	if !exists {
		return false, false
	}
	boolVal, ok := val.(bool)
	return boolVal, ok
}

// MarshalJSON for proper JSON encoding
func (t Tool) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Name        string                 `json:"name"`
		Description string                 `json:"description"`
		InputSchema map[string]interface{} `json:"input_schema"`
	}{
		Name:        t.Name,
		Description: t.Description,
		InputSchema: t.InputSchema,
	})
}
