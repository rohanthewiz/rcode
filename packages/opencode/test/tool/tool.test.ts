import { describe, expect, test } from "bun:test"
import { App } from "../../src/app/app"
import { GlobTool } from "../../src/tool/glob"
import { ListTool } from "../../src/tool/ls"

const ctx = {
  sessionID: "test",
  messageID: "",
  abort: AbortSignal.any([]),
  metadata: () => {},
}
describe("tool.glob", () => {
  test("truncate", async () => {
    await App.provide({ cwd: process.cwd() }, async () => {
      let result = await GlobTool.execute(
        {
          pattern: "../../node_modules/**/*",
          path: undefined,
        },
        ctx,
      )
      expect(result.metadata.truncated).toBe(true)
    })
  })
  test("basic", async () => {
    await App.provide({ cwd: process.cwd() }, async () => {
      let result = await GlobTool.execute(
        {
          pattern: "*.json",
          path: undefined,
        },
        ctx,
      )
      expect(result.metadata).toMatchObject({
        truncated: false,
        count: 3,
      })
    })
  })
})

describe("tool.ls", () => {
  test("basic", async () => {
    const result = await App.provide({ cwd: process.cwd() }, async () => {
      return await ListTool.execute(
        { path: "./example", ignore: [".git"] },
        ctx,
      )
    })
    expect(result.output).toMatchSnapshot()
  })
})
