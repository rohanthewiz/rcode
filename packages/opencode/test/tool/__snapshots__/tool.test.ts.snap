// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`tool.ls basic 1`] = `
"- /home/thdxr/dev/projects/sst/opencode/js/example/
  - home/
    - thdxr/
      - dev/
        - projects/
          - sst/
            - opencode/
              - js/
                - example/
                  - ink.tsx
                  - broken.ts
                  - cli.ts
"
`;
